### DISCUSSION TITLE
Reading multi-dimensional time data from a .nc file.

### DISCUSSION THREAD

**Discussion by sososmo** [NCAR/noahmp#162] [Ideas]
code

> !!!author?:soso
>  subroutine get_3d_netcdf(name, ncid, array, units, xstart, xend, ystart, yend,itime,  &
>        fatal_if_error, ierr)
>     implicit none
>     character(len=*), intent(in) :: name
>     integer, intent(in) :: ncid
>     integer, intent(in) :: itime
>     integer, intent(in) :: xstart, xend, ystart, yend
> #ifdef MPP_LAND
>     real, dimension(xstart:xend,ystart:yend), intent(inout) :: array
> #else
>     real, dimension(xstart:xend,ystart:yend), intent(out) :: array
> #endif
>     character(len=*), intent(out) :: units
>     integer :: iret, varid
>     integer :: itime_star
>     ! FATAL_IF_ERROR:  an input code value:
>     !      .TRUE. if an error in reading the data should stop the program.
>     !      Otherwise the, IERR error flag is set, but the program continues.
>     logical, intent(in) :: fatal_if_error 
>     integer, intent(out) :: ierr
> #ifdef MPP_LAND
>     real:: g_array(global_nx,global_ny)
> #endif
>     units = " "
>     
> #ifdef MPP_LAND
>     if(my_id .eq. 0) then
> #endif
> 
>     iret = nf90_inq_varid(ncid,  name,  varid)
>     if (iret /= 0) then
>        if (FATAL_IF_ERROR) then
>           print*, 'ncid = ', ncid
>           call error_handler(iret, "MODULE_HRLDAS_NETCDF_IO:  Problem finding variable '"//trim(name)//"' in NetCDF file.")
>        else
>           ierr = iret
> #ifdef MPP_LAND
>           goto 9991
> #endif
>           return
>        endif
>     endif
> 
>     iret = nf90_get_att(ncid, varid, "units", units)
>     if (iret /= 0) units = "units unknown"
> !KWM    if (iret /= 0) then
> !KWM       if (FATAL_IF_ERROR) then
> !KWM          print*, 'name = "', trim(name)//'"'
> !KWM          stop "MODULE_NOAHLSM_HRLDAS_INPUT:  get_2d_netcdf:  nf90_get_att:  units."
> !KWM       else
> !KWM          ierr = iret
> !KWM          return
> !KWM       endif
> !KWM    endif
>     itime_star = itime + 1
> #ifdef MPP_LAND
>     iret = nf90_get_var(ncid, varid, values=g_array, start=(/1,1,itime_star/), count=(/global_nx,global_ny,1/))
>     !iret = nf90_get_var(ncid, varid, values=g_array, start=(/itime_star,1,1/), count=(/1,global_nx,global_ny/))
>     !print*,'name_forc:',name
>     !if (name=='TAIR') then
>     !    print*,'g_array',g_array
>     !endif
> #else
>     iret = nf90_get_var(ncid, varid, values=array, start=(/xstart,ystart,itime_star/), count=(/xend-xstart+1,yend-ystart+1,1/))
>     !iret = nf90_get_var(ncid, varid, values=array, start=(/itime_star,xstart,ystart/), count=(/1,xend-xstart+1,yend-ystart+1/))
>     !    if (name=='TAIR') then
>     !    print*,'array:',array
>     !endif
> #endif
>     if (iret /= 0) then
> #ifdef MPP_LAND
>           goto 9991
> #endif
>        if (FATAL_IF_ERROR) then
>           print*, 'ncid =', ncid
>           call error_handler(iret, "MODULE_HRLDAS_NETCDF_IO:  Problem retrieving variable '"//trim(name)//"' from NetCDF file.")
>        else
>           ierr = iret
>           return
>        endif
>     endif
> 
>     ierr = 0;
> #ifdef MPP_LAND
> 9991    continue
>     endif
>     call mpp_land_bcast_int1(ierr)
>     !yyww call mpp_land_bcast(units)
>     call mpp_land_bcast_char(256,units)
>     if(ierr /= 0) return
>   !  write(6,*) "xstart,xend, ystart, yend", xstart,xend, ystart, yend
>     call decompose_data_real(g_array,array)
> #endif
>   end subroutine get_3d_netcdf

call
> call get_3d_netcdf(trim(forcing_name_T) , ncid, nextread%t,     units, xstart, xend, ystart, yend,time_dimentions, FATAL, ierr)

**Comment by cenlinhe** [comment]
Could you please add some descriptions for your proposed code? What is this code about? Is this a bug in the existing code you proposed to fix? Is this some new capabilities you would like to add?
